### PR TITLE
Move .su library code to TruffleRuby

### PR DESCRIPTION
**Pull request by eregon**
`SULONG_HOME` is only needed to get the options now:
```ruby
[
      "-J-Dpolyglot.llvm.libraryPath=#{SULONG_HOME}/mxbuild/sulong-libs",
      '-J-cp', "#{SULONG_HOME}/build/sulong.jar",
]
```
Both of these are automatically set in GraalVM, so we should now be able to compile and run C-extensions in GraalVM!

### CODE REVIEW

**Comment by eregon** [comment]
cc @grimmerm

**Comment by chrisseaton** [comment]
So now we don't need Sulong to build at all? But we've done that by copying the interface to Sulong (`truffle.h`) into our repo. Is that the right thing to do? I'm not entirely sure.

File: src/main/java/org/truffleruby/cext/CExtNodes.java [hunk 1]
```diff
@@ -1133,4 +1139,30 @@ public Object executeThrow(Throwable e,
...
+    public abstract static class LinkerNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization(guards = { "isRubyString(outputFileName)", "isRubyArray(libraries)", "isRubyArray(bitcodeFiles)" })
+        public Object linker(DynamicObject outputFileName, DynamicObject libraries, DynamicObject bitcodeFiles) {
+            try {
+                Linker.link(
+                        StringOperations.getString(outputFileName),
+                        array2StringList(libraries),
+                        array2StringList(bitcodeFiles));
+            } catch (NoSuchAlgorithmException | IOException e) {
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/cext/CExtNodes.java]: `NoSuchAlgorithmException`?

File: src/main/ruby/core/rbconfig.rb [hunk 2]
```diff
@@ -141,43 +141,42 @@ module RbConfig
...
+  mkconfig['COMPILE_C'] = "ruby #{libdir}/cext/preprocess.rb < $< | $(CC) $(INCFLAGS) $(CPPFLAGS) $(CFLAGS) $(COUTFLAG) -xc - -o $@ && #{opt} #{opt_passes} $@ -o $@",
+  expanded['LINK_SO'] = "#{ruby_launcher} -e Truffle::CExt::Linker.main -- -o $@ $(OBJS) #{libs}"
+  mkconfig['LINK_SO'] = "#{ruby_launcher} -e Truffle::CExt::Linker.main -- -o $@ $(OBJS) $(LIBS)"
+  expanded['TRY_LINK'] = "#{clang} $(src) $(INCFLAGS) #{cflags} #{libs}"
+  mkconfig['TRY_LINK'] = "#{clang} $(src) $(INCFLAGS) $(CFLAGS) $(LIBS)"
 
   def self.ruby
-    Truffle::Boot.ruby_launcher or
-        raise "we can't find the TruffleRuby launcher - set -Xlauncher=, or your launcher should be doing this for you"
+    Truffle::Boot.ruby_launcher
```
Comment (chrisseaton) [inline src/main/ruby/core/rbconfig.rb]: Why this change?

File: src/main/ruby/core/truffle/cext.rb [hunk 3]
```diff
@@ -15,6 +15,36 @@ class Data
...
+          else
+            files << arg
+          end
+        end
+      end
+
+      Truffle::CExt.linker(output, libraries, files)
+    end
+  end
+
```
Comment (chrisseaton) [inline src/main/ruby/core/truffle/cext.rb]: I think we should probably have a launcher for this - but we can worry about that later when we look at users compiling their own C extensions.

File: src/main/ruby/core/rbconfig.rb (hunk 2 above)
Comment (eregon) [inline src/main/ruby/core/rbconfig.rb]: It's checked above, but apparently we still have many configurations which do not have the launcher set :(

File: src/main/java/org/truffleruby/cext/CExtNodes.java (hunk 1 above)
Comment (eregon) [inline src/main/java/org/truffleruby/cext/CExtNodes.java]: It's what `link` throws.

File: src/main/ruby/core/truffle/cext.rb (hunk 3 above)
Comment (eregon) [inline src/main/ruby/core/truffle/cext.rb]: `truffleruby -e Truffle::CExt::Linker.main -- *args` seems a nice way to avoid an extra launcher (see RbConfig above).

File: src/main/java/org/truffleruby/cext/CExtNodes.java (hunk 1 above)
Comment (eregon) [inline src/main/java/org/truffleruby/cext/CExtNodes.java]: Fixed in f1e992cb7535f081df6d5b34f58338f18a571b93

**Comment by chrisseaton** [comment]
Yes we don't always have a launcher - which makes sense for embedded cases. You'll need to work around that.

**Comment by eregon** [comment]
OK, I allowed not having a launcher again.

> So now we don't need Sulong to build at all? But we've done that by copying the interface to Sulong (truffle.h) into our repo. Is that the right thing to do? I'm not entirely sure.

Yes, and we only need to be able to interop eval to use Sulong.
Copying truffle.h should be fine, it's expected to stay compatible, it's small and we can just copy it over when we need a new function. I discussed it with the Sulong team.
